File: app/src/model/viz_spec.ts
import { TopLevelSpec } from 'vega-lite/build/src/spec/index.js';

export interface VizSpec {
    readonly renderer: CardRendererData;
}

export type CardRendererData = TableRendererData | VegaLiteRendererData;

export interface TableRendererData {
    readonly t: 'Table';
    readonly v: {
        table_name: string;
        row_count?: number;
    };
}

export interface VegaLiteRendererData {
    readonly t: 'VegaLite';
    readonly v: {
        table_name: string;
        sampling?: SamplingMethod;
        spec: TopLevelSpec;
    };
}

export type SamplingMethod = ReservoirSampling | AM4Sampling;

export interface ReservoirSampling {
    readonly t: 'Reservoir';
    readonly v: number;
}

export interface AM4Sampling {
    readonly t: 'AM4';
    readonly v: {
        attribute_x: string;
        attribute_y: string;
        domain_x: any[];
    };
}

File: libs/dashql-core/src/analyzer/viz_spec.rs
use std::sync::Arc;

use serde::Serialize;

use crate::execution::table_metadata::TableMetadata;

#[derive(Default, Debug, Clone, Serialize, PartialEq)]
pub struct VizSpec {
    pub renderer: VizRendererData,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(tag = "t", content = "v")]
pub enum VizRendererData {
    Table(TableRendererData),
    VegaLite(VegaLiteRendererData),
    Json(JsonRendererData),
    Hex(HexRendererData),
}

#[derive(Default, Debug, Clone, Serialize, PartialEq)]
pub struct TableRendererData {
    pub table_name: String,
    pub table_metadata: Arc<TableMetadata>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct VegaLiteRendererData {
    pub table_name: String,
    pub table_metadata: Arc<TableMetadata>,
    pub sampling: Option<SamplingMethod>,
    pub spec: serde_json::Map<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct JsonRendererData {
    pub source_data_id: u32,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct HexRendererData {
    pub source_data_id: u32,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(tag = "t", content = "v")]
pub enum SamplingMethod {
    Reservoir(u32),
    AM4(AM4Config),
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(tag = "t", content = "v")]
pub enum DomainValue {
    Null,
    String(String),
    F64(f64),
    Bool(bool),
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AM4Config {
    pub attribute_x: String,
    pub attribute_y: String,
    pub domain_x: Vec<DomainValue>,
}

impl Default for VizRendererData {
    fn default() -> Self {
        VizRendererData::Table(TableRendererData::default())
    }
}
